File: src/main/java/com/example/flightradar/AircraftPoller.java
package com.example.flightradar;

import java.util.List;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

/**
 * Periodically polls an {@link OpenSkyClient} on a background thread and delivers each batch of
 * aircraft states to a callback.
 *
 * <p>The callback runs on the poller's own (non-UI) thread. Callers that touch JavaFX must marshal
 * back onto the JavaFX Application Thread themselves (e.g. via {@code Platform.runLater}).
 */
public final class AircraftPoller {

    private final OpenSkyClient client;
    private final long periodSeconds;
    private final Consumer<List<AircraftState>> onUpdate;

    private ScheduledExecutorService scheduler;

    public AircraftPoller(OpenSkyClient client, long periodSeconds,
                          Consumer<List<AircraftState>> onUpdate) {
        this.client = client;
        this.periodSeconds = periodSeconds;
        this.onUpdate = onUpdate;
    }

    /** Starts polling immediately, then every {@code periodSeconds}. Idempotent-safe per instance. */
    public synchronized void start() {
        if (scheduler != null) {
            return;
        }
        scheduler = Executors.newSingleThreadScheduledExecutor(runnable -> {
            Thread thread = new Thread(runnable, "aircraft-poller");
            thread.setDaemon(true);
            return thread;
        });
        scheduler.scheduleAtFixedRate(this::poll, 0, periodSeconds, TimeUnit.SECONDS);
    }

    private void poll() {
        try {
            List<AircraftState> states = client.fetchStates();
            onUpdate.accept(states);
        } catch (Exception e) {
            // Never let an exception kill the scheduled task.
            System.err.println("[Poller] update failed: " + e.getMessage());
        }
    }

    /** Stops polling and releases the background thread. */
    public synchronized void stop() {
        if (scheduler != null) {
            scheduler.shutdownNow();
            scheduler = null;
        }
    }
}

File: python/flightradar/aircraft_poller.py
"""Periodically polls an :class:`OpenSkyClient` on a background thread and delivers each batch of
aircraft states to a callback.

The callback runs on the poller's own (non-UI) thread. Callers that touch Tkinter must marshal back
onto the Tk main thread themselves (e.g. via ``root.after(0, ...)``).
"""

import threading


class AircraftPoller:
    def __init__(self, client, period_seconds, on_update):
        self._client = client
        self._period = period_seconds
        self._on_update = on_update
        self._stop = threading.Event()
        self._thread = None

    def start(self):
        """Start polling immediately, then every ``period_seconds``. Idempotent per instance."""
        if self._thread is not None:
            return
        self._thread = threading.Thread(target=self._run, name="aircraft-poller", daemon=True)
        self._thread.start()

    def _run(self):
        while not self._stop.is_set():
            try:
                states = self._client.fetch_states()
                self._on_update(states)
            except Exception as exc:  # noqa: BLE001 - never let an exception kill the loop
                print(f"[Poller] update failed: {exc}")
            self._stop.wait(self._period)

    def stop(self):
        """Stop polling and release the background thread."""
        self._stop.set()
